### PR TITLE
Add Seed admin dashboard page

### LINKED ISSUE #187: Add Seed Admin Dashboard Page
@BartuCulha This is part of: https://github.com/PrimeDAO/prime-launch-dapp/issues/208

## User Story
As an admin I would like to access and use the admin functionality provided by the contracts, inside the Prime Launch Dapp.

## Current Scenario
The Prime Launch Dapp Doesn't have the Admin Dashboard as of now.

## Expected Result
The Prime Launch Dapp should have minimalist page for Admin to access the following functionality provided by the contracts. 
- [ ] Add Whitelist
- [ ] Add an address to Whitelist
- [ ] Remove an address from whitelist
- [ ] Retrieve Project Tokens
- [ ] Withdraw funding tokens
- [ ] Pause/Unpause
- [ ] Close
- [ ] Easy way to fund Seed with required amount of project tokens.

## Acceptance Criteria
- [ ] Implement all the functionality
- [ ] Minimalist UI
- [ ] According to the wireframe
- [ ] Well structured Code

### PR DESCRIPTION
**Pull request by codebuster22**
closes #187

### CODE REVIEW

File: src/adminDashboard/adminDashboard.html [hunk 1]
```diff
@@ -3,8 +3,78 @@
...
+                <div class="stageButton proceed button1" click.delegate="closeLaunch()">
+                  Close
+                </div>
+              </div>
+              <div>
+                <div>
+                  <label for="addressToRemove">Address to remove from whitelist</label>
+                  <input id="addressToRemove" value.bind="addressToRemove" placeholder="0x6C4ef..." />
+                </div>
+                <div class="stageButton proceed button1" click.delegate="removeFromWhiteliste(addressToRemove)">
```
Comment (dkent600) [inline src/adminDashboard/adminDashboard.html:52]: typo in function name

File: src/adminDashboard/adminDashboard.ts [hunk 2]
```diff
@@ -1,17 +1,23 @@
-import { EthereumService } from "services/EthereumService";
-import { Seed } from "entities/Seed";
+import BigNumber from '../services/BigNumberService';
```
Comment (dkent600) [inline src/adminDashboard/adminDashboard.ts:1]: Why is this here?  In any case, unless under certain circumstances, you should use the ethers.js version of BigNumber, not this service.

File: src/adminDashboard/adminDashboard.ts [hunk 3]
```diff
@@ -51,14 +57,100 @@ export class AdminDashboard {
   async hydrate(): Promise<void> {
     if (this.ethereumService.defaultAccountAddress) {
       const defaultAccount = this.ethereumService.defaultAccountAddress.toLowerCase();
-      this.seeds = this.seedService.seedsArray
-        .filter((seed) => { return seed.admin.toLowerCase() === defaultAccount;});
+      this.seeds = this.seedService.seedsArray;
```
Comment (dkent600) [inline src/adminDashboard/adminDashboard.ts:60]: why aren't you filtering by defaultAccount?

File: src/adminDashboard/adminDashboard.ts [hunk 4]
```diff
@@ -51,14 +57,100 @@ export class AdminDashboard {
...
+      await this.selectedSeed.contract.withdraw();
+    } catch (error) {
+      alert(error.message);
+    }
+  }
+
+  async addWhitelist() {
+    try {
+      const filterPattern = /([^\W]+)/g;
+      const response = await axios.get(this.selectedSeed.metadata.seedDetails.whitelist);
```
Comment (dkent600) [inline src/adminDashboard/adminDashboard.ts:99]: please use our WhitelistService for fetching, adding a list, and adding and removing addresses.

File: src/adminDashboard/adminDashboard.ts [hunk 5]
```diff
@@ -51,14 +57,100 @@ export class AdminDashboard {
...
+      await this.selectedSeed.contract.unwhitelist(address);
+    } catch (error) {
+      alert(error.message);
+    }
+  }
+
+  async pauseLaunch() {
+    try {
+      await this.selectedSeed.contract.callStatic.pause();
+      await this.selectedSeed.contract.pause();
```
Comment (dkent600) [inline src/adminDashboard/adminDashboard.ts:129]: I think it would be better if you add all these Seed functions to the Seed entity.

File: src/adminDashboard/adminDashboard.ts [hunk 6]
```diff
@@ -51,14 +57,100 @@ export class AdminDashboard {
...
   }
 
-  connect():void {
+  async setSeed(index) {
+    this.selectedSeed = this.seeds[index];
+  }
+
+  async fundSeed() {
+    try {
+      await this.selectedSeed.projectTokenContract.transfer(this.selectedSeed.address, this.selectedSeed.seedAmountRequired);
```
Comment (dkent600) [inline src/adminDashboard/adminDashboard.ts:72]: You should be using the TransactionsService for sending any transaction.

**Review by dkent600 — CHANGES_REQUESTED** [review]

File: src/adminDashboard/adminDashboard.ts (hunk 3 above)
Comment (codebuster22) [reply src/adminDashboard/adminDashboard.ts:60]: Because I'm not an admin of any seed, and to test this, I wasn't filtering it. I'll add filtering once I create a seed.

File: src/adminDashboard/adminDashboard.ts (hunk 5 above)
Comment (codebuster22) [reply src/adminDashboard/adminDashboard.ts:129]: by Seed entity we mean, the SeedService or this.selectedSeed?

File: src/adminDashboard/adminDashboard.ts (hunk 6 above)
Comment (codebuster22) [reply src/adminDashboard/adminDashboard.ts:72]: okay

File: src/adminDashboard/adminDashboard.html [hunk 7]
```diff
@@ -3,8 +3,78 @@
     <div show.bind="!loading">
       <div if.bind="!connected"><button class="button1" click.delegate="connect()">Connect...</button></div>
       <div else class="formContainer">
-        <div if.bind="seeds.length">[Seeds Dropdown, etc...]</div>
-        <div else>You have not created any Seeds</div>
+        <div>
```
[the remark this answers is not in the document]
Comment (codebuster22) [reply src/adminDashboard/adminDashboard.html:6]: done

File: src/adminDashboard/adminDashboard.html [hunk 8]
```diff
@@ -3,8 +3,78 @@
...
+          <div if.bind="seeds.length">
+            <div>
+              <dropdown item-changed.call="setSeed(index)"
+                selected-item-index.to-view="selectedSeed.metadata.general.projectName"
+                placeholder="Select a Category...">
```
[the remark this answers is not in the document]
Comment (codebuster22) [reply src/adminDashboard/adminDashboard.html:11]: done

File: src/adminDashboard/adminDashboard.html [hunk 9]
```diff
@@ -3,8 +3,78 @@
...
+                <div repeat.for="seed of seeds">${seed.metadata.general.projectName}</div>
+              </dropdown>
+            </div>
+            <div if.bind="selectedSeed.address">
+              Project Name:- ${selectedSeed.metadata.general.projectName}
```
[the remark this answers is not in the document]
Comment (codebuster22) [reply src/adminDashboard/adminDashboard.html:16]: Removed this

File: src/adminDashboard/adminDashboard.html [hunk 10]
```diff
@@ -3,8 +3,78 @@
...
+              <div>
+                <div class="stageButton proceed button1" click.delegate="fundSeed()">
+                  Fund Project
+                </div>
+              </div>
+              <div>
```
[the remark this answers is not in the document]
Comment (codebuster22) [reply src/adminDashboard/adminDashboard.html:22]: I added Div to have it in separate line or container. I have removed it though.

File: src/adminDashboard/adminDashboard.html (hunk 1 above)
Comment (codebuster22) [reply src/adminDashboard/adminDashboard.html:52]: Resolved the typo

File: src/adminDashboard/adminDashboard.ts (hunk 2 above)
Comment (codebuster22) [reply src/adminDashboard/adminDashboard.ts:1]: done, removed the unnecessary imports

Comment (codebuster22) [reply src/adminDashboard/adminDashboard.ts:1]: Will make sure to use ethers.js

File: src/adminDashboard/adminDashboard.ts (hunk 4 above)
Comment (codebuster22) [reply src/adminDashboard/adminDashboard.ts:99]: done

File: src/adminDashboard/adminDashboard.ts (hunk 5 above)
Comment (codebuster22) [reply src/adminDashboard/adminDashboard.ts]: I have added to this file `src/entities/Seed.ts`.

File: src/adminDashboard/adminDashboard.ts (hunk 6 above)
Comment (codebuster22) [reply src/adminDashboard/adminDashboard.ts]: done

File: src/adminDashboard/adminDashboard.html [hunk 11]
```diff
@@ -3,8 +3,78 @@
...
+              <dropdown item-changed.call="setSeed(index)"
+                selected-item-index.to-view="selectedSeed.metadata.general.projectName"
+                placeholder="Select a Category...">
+                <div repeat.for="seed of seeds">${seed.metadata.general.projectName}</div>
+              </dropdown>
+            </div>
+            <div if.bind="selectedSeed.address">
+              Project Name:- ${selectedSeed.metadata.general.projectName}
+              <div>
+                <div class="stageButton proceed button1" click.delegate="fundSeed()">
```
[the remark this answers is not in the document]
Comment (codebuster22) [reply src/adminDashboard/adminDashboard.html]: done, removed irrelevant class.

File: src/adminDashboard/adminDashboard.html [hunk 12]
```diff
@@ -3,8 +3,61 @@
     <div show.bind="!loading">
       <div if.bind="!connected"><button class="button1" click.delegate="connect()">Connect...</button></div>
       <div else class="formContainer">
-        <div if.bind="seeds.length">[Seeds Dropdown, etc...]</div>
-        <div else>You have not created any Seeds</div>
+          <div if.bind="seeds.length">
+            <div>
```
Comment (dkent600) [inline src/adminDashboard/adminDashboard.html:7]: can you delete this div?

File: src/entities/Seed.ts [hunk 13]
```diff
@@ -396,6 +402,54 @@ export class Seed {
...
+    );
+  }
+
+  async removeFromWhitelist(address: Address): Promise<TransactionReceipt> {
+    return this.transactionsService.send(
+      () => this.contract.unwhitelist(address)
+    );
+  }
+
+  async pauseLaunch (): Promise<TransactionReceipt> {
```
Comment (dkent600) [inline src/entities/Seed.ts:435]: please just name this "pause"

File: src/entities/Seed.ts [hunk 14]
```diff
@@ -396,6 +402,54 @@ export class Seed {
...
+    return this.transactionsService.send(
+      () => this.contract.pause()
+    );
+  }
+
+  async unpauseLaunch(): Promise<TransactionReceipt> {
```
Comment (dkent600) [inline src/entities/Seed.ts:441]: please just name this "unPause"

File: src/entities/Seed.ts [hunk 15]
```diff
@@ -396,6 +402,54 @@ export class Seed {
...
+    return this.transactionsService.send(
+      () => this.contract.unpause()
+    );
+  }
+
+  async closeLaunch(): Promise<TransactionReceipt> {
```
Comment (dkent600) [inline src/entities/Seed.ts:447]: please just name this "close"

File: src/entities/Seed.ts [hunk 16]
```diff
@@ -374,6 +374,12 @@ export class Seed {
     }
   }
 
+  async fundSeed(): Promise<TransactionReceipt> {
```
Comment (dkent600) [inline src/entities/Seed.ts:377]: please just name this "fund"

File: src/entities/Seed.ts [hunk 17]
```diff
@@ -374,6 +374,12 @@ export class Seed {
...
+    return this.transactionsService.send(
+      () => this.projectTokenContract.transfer(this.contract.address, this.seedAmountRequired)
```
Comment (dkent600) [inline src/entities/Seed.ts:379]: for all of these new methods, can you please be sure you update the state of the seed similar to what you see being done in the `buy` method?  Note: don't hydrate user info unless needed.

**Review by dkent600 — CHANGES_REQUESTED** [review]

File: src/adminDashboard/adminDashboard.html [hunk 18]
```diff
@@ -3,8 +3,61 @@
...
+                <div if.bind="selectedSeed.hasEnoughProjectTokens" class="button1" click.delegate="selectedSeed.fundSeed()">
+                  Fund Project
+                </div>
+                <div class="button1" click.delegate="selectedSeed.withdrawFundingTokens()">
+                  Withdraw Funding Tokens
+                </div>
+                <div if.bind="selectedSeed.whitelisted" class="button1" click.delegate="addWhitelist()">
+                  Add Whitelist
+                </div>
+                <div if.bind="!selectedSeed.isPaused" class="button1" click.delegate="selectedSeed.pauseLaunch()">
```
Comment (dkent600) [inline src/adminDashboard/adminDashboard.html:24]: Please be sure you are checking for possible reverts for all of these functions, such as here, where maybe one can't pause if the seed is closed.

File: src/entities/Seed.ts (hunk 17 above)
Comment (codebuster22) [reply src/entities/Seed.ts:379]: okay, I'll do that.

File: src/adminDashboard/adminDashboard.html (hunk 18 above)
Comment (codebuster22) [reply src/adminDashboard/adminDashboard.html:24]: ohk, I'll update

File: src/entities/Seed.ts (hunk 17 above)
Comment (codebuster22) [reply src/entities/Seed.ts:379]: have done, but I have some doubts regarding how to update state (Have updated the code).
1. for `pause`, `unPause` and `close`, I should use `hydateClosedOrPaused` method provided in Seed entity, right?
2. for `retrieveProjectTokens` and `withdrawFundingTokens`, I should use `hydrateTokensState` method provided in Seed entity, right?
3. for functions related to whitelisting, should I update the state? sometime, admin might later add himself/herself to the whitelist, do we want to update the user info too?

File: src/adminDashboard/adminDashboard.html (hunk 18 above)
Comment (codebuster22) [reply src/adminDashboard/adminDashboard.html]: done, but have some doubts, regarding the `if.bind()`, I wanted to use multiple conditions to check for `withdraw` and `retrieveProjectTokens`, I have commented the conditions I would use, condition then be simplified though.
Using this conditions, I'm facing the issue, I think it's not able to recognize `||`.

File: src/entities/Seed.ts (hunk 16 above)
Comment (codebuster22) [reply src/entities/Seed.ts]: done

File: src/entities/Seed.ts (hunk 15 above)
Comment (codebuster22) [reply src/entities/Seed.ts]: done

File: src/entities/Seed.ts (hunk 14 above)
Comment (codebuster22) [reply src/entities/Seed.ts]: done

File: src/adminDashboard/adminDashboard.html (hunk 12 above)
Comment (codebuster22) [reply src/adminDashboard/adminDashboard.html]: done

File: src/adminDashboard/adminDashboard.ts (hunk 3 above)
Comment (codebuster22) [reply src/adminDashboard/adminDashboard.ts:63]: done

File: src/adminDashboard/adminDashboard.html (hunk 18 above)
Comment (codebuster22) [reply src/adminDashboard/adminDashboard.html]: done

File: src/entities/Seed.ts (hunk 17 above)
Comment (codebuster22) [reply src/entities/Seed.ts:386]: I think it's done. Can you check it please.

File: src/entities/Seed.ts [hunk 19]
```diff
@@ -403,6 +409,69 @@ export class Seed {
       });
   }
 
+  async addWhitelist(whitelistAddress: Set<Address>): Promise<TransactionReceipt> {
+    return this.transactionsService.send(
+      () => this.contract.whitelistBatch([...whitelistAddress])
+    ).then((receipt) => {
+      if(receipt){
+        this.hydrateUser()
+        return receipt
```
Comment (dkent600) [inline src/entities/Seed.ts:418]: please make sure you run lint on your files.

File: src/adminDashboard/adminDashboard.ts [hunk 20]
```diff
@@ -1,19 +1,26 @@
...
 export class AdminDashboard {
 
   seeds: Array<Seed> = [];
+  selectedSeed: Seed;
+  addressToRemove: String = '';
+  addressToAdd: String = '';
+  receiverAddress: String = '';
   subscriptions: DisposableCollection = new DisposableCollection();
-  loading = true;
+  loading: boolean = true;
```
Comment (dkent600) [inline src/adminDashboard/adminDashboard.ts:23]: just FYI, `loading = true` is fine because typescript infers the datatype from the assignment.  but definitely doesn't hurt to be explicit.

**Review by dkent600 — CHANGES_REQUESTED** [review]
I think we need to rename these files to specify they are for seeds, like "seedAdminDashboard"

File: src/entities/Seed.ts (hunk 19 above)
Comment (codebuster22) [reply src/entities/Seed.ts:418]: 👍

File: src/adminDashboard/adminDashboard.ts (hunk 20 above)
Comment (codebuster22) [reply src/adminDashboard/adminDashboard.ts]: ohk. I haven't changed it, should I remove the type `boolean`

File: src/entities/Seed.ts (hunk 19 above)
Comment (codebuster22) [reply src/entities/Seed.ts]: done

**Review by dkent600 — CHANGES_REQUESTED** [review]
Looks good to move to the UI

**Review by dkent600 — CHANGES_REQUESTED** [review]
Needs to follow the mock given here:

https://www.figma.com/file/zRYES2gtCKH59mv7cHqSeH/Admin-Page?node-id=165%3A542

1) but remove the line under the header.  this is inconsistent with our design elsewhere.
2) Connect button should be under the header, not above.  Should be some grey text that says "Connect your wallet to administer your Seeds")
3) The dropdown should not be visible if there is only one seed
4) respect the line spacings that Bartu has given in the mock
5) header should be the gradient colors

**Review by dkent600 — CHANGES_REQUESTED** [review]
This should show the page header and the message text should be centered

image

There is not header even with seeds:

image

Widths don't align.  Talk to Gilad about this:

image
